operator/builtin/input/file/file_test.go

```diff
@@ -681,65 +686,182 @@ func TestMultiCopyTruncateSlow(t *testing.T) {
+	maxBackupsExceeded := rt.totalLines > maxLinesInAllFiles
+
+	// last line written in primary file will exist for l*b more writes
+	minTimeToLive := time.Duration(int(rt.writeInterval) * rt.maxLinesPerFile * rt.maxBackupFiles)
+
+	// can a line be written and then rotated to deletion before ever observed?
+	return maxBackupsExceeded && rt.pollInterval > minTimeToLive
+}
 
-	for i := 0; i < numMessages; i++ {
-		expected = append(expected, getMessage(i))
-	}
+type rotationTest struct {
+	name            string
+	totalLines      int
+	maxLinesPerFile int
+	maxBackupFiles  int
+	writeInterval   time.Duration
+	pollInterval    time.Duration
+	ephemeralLines  bool
+}
 
-	require.NoError(t, operator.Start())
-	defer operator.Stop()
+func TestRotation(t *testing.T) {
 
-	for _, message := range expected {
-		logger.Writer().Write([]byte(message + "\n"))
-		time.Sleep(200 * time.Microsecond)
+	cases := []rotationTest{
+		{
+			name:            "Fast/NoRotation",
+			totalLines:      10,
+			maxLinesPerFile: 10,
+			maxBackupFiles:  1,
+			writeInterval:   time.Millisecond,
+			pollInterval:    20 * time.Millisecond,
+		},
+		{
+			name:            "Fast/NoDeletion",
+			totalLines:      20,
+			maxLinesPerFile: 10,
+			maxBackupFiles:  1,
+			writeInterval:   time.Millisecond,
+			pollInterval:    20 * time.Millisecond,
+		},
+		{
+			name:            "Fast/Deletion",
+			totalLines:      30,
+			maxLinesPerFile: 10,
+			maxBackupFiles:  1,
+			writeInterval:   time.Millisecond,
+			pollInterval:    20 * time.Millisecond,
+			ephemeralLines:  true,
+		},
+		{
+			name:            "Fast/Deletion/ExceedFingerprint",
+			totalLines:      300,
+			maxLinesPerFile: 100,
+			maxBackupFiles:  1,
+			writeInterval:   time.Millisecond,
+			pollInterval:    20 * time.Millisecond,
+			ephemeralLines:  true,
+		},
+		{
+			name:            "Slow/NoRotation",
+			totalLines:      10,
+			maxLinesPerFile: 10,
+			maxBackupFiles:  1,
+			writeInterval:   3 * time.Millisecond,
+			pollInterval:    20 * time.Millisecond,
+		},
+		{
+			name:            "Slow/NoDeletion",
+			totalLines:      20,
+			maxLinesPerFile: 10,
+			maxBackupFiles:  1,
+			writeInterval:   3 * time.Millisecond,
+			pollInterval:    20 * time.Millisecond,
+		},
+		{
+			name:            "Slow/Deletion",
+			totalLines:      30,
+			maxLinesPerFile: 10,
+			maxBackupFiles:  1,
+			writeInterval:   3 * time.Millisecond,
+			pollInterval:    20 * time.Millisecond,
+		},
+		{
+			name:            "Slow/Deletion/ExceedFingerprint",
+			totalLines:      300,
+			maxLinesPerFile: 100,
+			maxBackupFiles:  1,
+			writeInterval:   3 * time.Millisecond,
+			pollInterval:    20 * time.Millisecond,
+		},
 	}
 
-	waitForMessages(t, logReceived, expected)
-}
-
-func TestRapidRotate_CopyTruncate(t *testing.T) {
-	t.Skip()
-	getMessage := func(m int) string { return fmt.Sprintf("message %d", m) }
-
-	numMessages := 1000
+	testRun := func(tc rotationTest, copyTruncate bool) func(t *testing.T) {
```

Comment:
Could we make this a method of the rotation tests rather than a closure?

Reply:
Great idea. Will do